dbos/workflow.go

```diff
@@ -538,45 +526,62 @@ func (c *dbosContext) RunAsWorkflow(_ DBOSContext, fn WorkflowFunc, input any, o
 		status = WorkflowStatusPending
 	}
 
+	// Check if the user-provided context has a deadline
+	deadline, ok := c.Deadline()
+	if !ok {
+		deadline = time.Time{} // No deadline set
+	}
+
+	// Compute the timeout based on the deadline
+	var timeout time.Duration
+	if !deadline.IsZero() {
+		timeout = time.Until(deadline)
+		/* unclear to me if this is a real use case:
+		if timeout < 0 {
+			return nil, newWorkflowExecutionError(workflowID, "deadline is in the past")
+		}
+		*/
```

Comment:
`timeout < 0` if the deadline is in the past. In this case the cancellation logic should already have been started. I might do something like checking the cancellation map if we hit that case -- we'll see.